### PR TITLE
Create LatestCheckStatusCondition

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

This creates the first AutomationCondition that operates directly on checks. It allows users to incorporate the status of the latest check that targeted the current version of the asset.

Two larger changes worth discussing:

1. I added an `all_deps_blocking_checks_succeeded` pre-built condition which should solve the long-standing user request of making "AMP" not materialize assets in the case that upstream blocking checks have failed
2. For check status conditions, I went with a syntax of `AutomationCondition.check.<status>`. I agonized about alternatives for awhile, with the main sticking point being that it means slightly different things for an asset to fail / succeed / be in progress than it means for an asset check to be in one of those statuses. Thus, we need to distinguish between `AutomationCondition.in_progress` (which is asset-specific) and `AutomationCondition.check.in_progress`. I thought `AutomationCondition.check_in_progress` felt a little off, almost like it's saying "A check on an asset is currently in progress" rather than "This is a check that is currently in progress."

Very open to alternatives though.

## Changelog

NOCHANGELOG

- [ ] `NEW` _(added new feature or capability)_
- [ ] `BUGFIX` _(fixed a bug)_
- [ ] `DOCS` _(added or updated documentation)_

### CODE REVIEW

**Comment by OwenKephart** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24384?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#24384** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24384?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#24359** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24359?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#24368](https://github.com/dagster-io/dagster/pull/24368) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24368?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#24356** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24356?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24283** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24283?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#24285](https://github.com/dagster-io/dagster/pull/24285) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24285?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#24282** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24282?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24276** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24276?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24323** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24323?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24241** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24241?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24202** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24202?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#24237](https://github.com/dagster-io/dagster/pull/24237) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24237?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#24200** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24200?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24111** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24111?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23910** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23910?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23907** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23907?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23905** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23905?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23904** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23904?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23843** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23843?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#23842](https://github.com/dagster-io/dagster/pull/23842) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23842?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by OwenKephart** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24384?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#24384** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24384?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#24359** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24359?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#24368](https://github.com/dagster-io/dagster/pull/24368) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24368?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#24356** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24356?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24283** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24283?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#24285](https://github.com/dagster-io/dagster/pull/24285) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24285?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#24282** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24282?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24276** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24276?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24323** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24323?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24241** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24241?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24202** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24202?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#24237](https://github.com/dagster-io/dagster/pull/24237) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24237?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#24200** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24200?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24111** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24111?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23910** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23910?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23907** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23907?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23905** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23905?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23904** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23904?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23843** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23843?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#23842](https://github.com/dagster-io/dagster/pull/23842) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23842?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/base_asset_graph.py [hunk 1]
```diff
@@ -305,10 +305,11 @@ def toposorted_entity_keys(self) -> Sequence[EntityKey]:
         """Return topologically sorted entity keys in graph. Keys with the same topological level are
         sorted alphabetically to provide stability.
         """
+        sort_key = lambda e: str(e)
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/base_asset_graph.py:308]: this is annoying -- needed to do this because you can't natively sort a list of `Union[AssetKey, AssetCheckKey]`.

File: python_modules/dagster/dagster_tests/definitions_tests/declarative_automation_tests/automation_condition_tests/builtins/test_check_status_condition.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    "status, automation_condition",
+    [
+        (ACS.SUCCEEDED, AutomationCondition.check.succeeded()),
+        (ACS.FAILED, AutomationCondition.check.failed()),
+        (ACS.EXECUTION_FAILED, AutomationCondition.check.execution_failed()),
+    ],
+)
+def test_status_conditions(status: ACS, automation_condition: AutomationCondition) -> None:
+    defs = _get_defs(status, automation_condition)
+    asset_job = defs.get_implicit_global_asset_job_def()
```
Comment (schrockn) [inline python_modules/dagster/dagster_tests/definitions_tests/declarative_automation_tests/automation_condition_tests/builtins/test_check_status_condition.py:49]: IMO these type of. parameterizations often hurt more than they help as they are hard to parse. Would rather seem some more straightforward examples that more clearly mimic what the users would write.

File: python_modules/dagster/dagster_tests/definitions_tests/declarative_automation_tests/automation_condition_tests/builtins/test_check_status_condition.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    @asset(
+        check_specs=[
+            AssetCheckSpec(name="b", asset="A", blocking=True),
+            AssetCheckSpec(name="nb", asset="A", blocking=False),
+        ]
+    )
+    def A(context: AssetExecutionContext):
+        yield AssetCheckResult(check_name="b", passed="a_fail_b" not in context.run_tags)
+        yield AssetCheckResult(check_name="nb", passed="a_fail_nb" not in context.run_tags)
+        yield Output(None)
```
Comment (schrockn) [inline python_modules/dagster/dagster_tests/definitions_tests/declarative_automation_tests/automation_condition_tests/builtins/test_check_status_condition.py:85]: would prefer to use `MaterializeResult` when we aren't using I/O managers etc

File: python_modules/dagster/dagster_tests/definitions_tests/declarative_automation_tests/automation_condition_tests/builtins/test_check_status_condition.py (hunk 2 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster_tests/definitions_tests/declarative_automation_tests/automation_condition_tests/builtins/test_check_status_condition.py]: I generally agree, and I think the last test here does a better job of illustrating a "real-world" use case (at the expense of being less targeted).

There's a bit of inherent weirdness here with testing a single automation condition in a vacuum, as many of these "basic" automation conditions would not make sense as a standalone condition on an asset.

I'd prefer to keep this specific test around (as it does capture some valuable cases), but I also added an additional "more realistic" test underneath.

**Review by schrockn — CHANGES_REQUESTED** [review]
I've been thinking about this and the `AutomationCondition.check` thing is nagging at me and is kind of a tell that our vocabulary is not sufficiently precise. The lack of formal separation between the entity in the graph and the underlying compute that backs it is really killing us here.

Certainly as a user I would expect `AutomationCondition.in_progress` to apply to either noun. 

In terms of vocabulary we do distinguish "pass" from "succeed" which gives us some room to maneuver I think.

Additionally I actually think `check.failed` is its own footgun here as currently proposed. Does it also count as failed if the latest _execution_ failed but did not set a check status? Either way its a big ambiguous.

What are all the alternatives that you thought of?

Browsing the docblocks one possibility is to make the notion of "runs" more explicitly:

* Change `failed` to `latest_run_failed` for all types. I think this might be clearer anyway and disambiguates from any notion of future asset health or other concepts.
* `check_passed` and `check_failed` become much clearer then.

Note: `in_progress` docblock should be clarified re backfills versus runs.
